src/AppInstallerRepositoryCore/PackageDependenciesValidation.cpp

```diff
@@ -0,0 +1,253 @@
+
+            if (results.Matches.empty())
+            {
+                return result;
+            }
+
+            auto packageRowId = results.Matches[0].first;
+
+            auto vac = index->GetVersionKeysById(packageRowId);
+
+            Utility::Version maxVersion = Utility::Version::CreateUnknown();
+
+            for (auto& v : vac)
+            {
+                auto currentVersion = v.GetVersion();
+                if (exclusions.find(currentVersion) != exclusions.end())
+                {
+                    continue;
+                }
+
+                if (maxVersion.IsUnknown() || currentVersion > maxVersion)
+                {
+                    maxVersion = currentVersion;
+                }
+            }
+
+            if (!maxVersion.IsUnknown())
+            {
+                auto manifestRowId = index->GetManifestIdByKey(packageRowId, maxVersion.ToString(), "");
+                result = std::make_pair(manifestRowId.value(), maxVersion);
+            }
+
+            return result;
+        }
+    };
+
+    bool PackageDependenciesValidation::ValidateManifestDependencies(SQLiteIndex* index, const Manifest::Manifest manifest)
+    {
+        using namespace Manifest;
+
+        Dependency rootId(DependencyType::Package, manifest.Id, manifest.Version);
+        std::vector<ValidationError> dependenciesError;
+        bool foundErrors = false;
+
+        DependencyGraph graph(rootId, [&](const Dependency& node) {
+
+            DependencyList depList;
+            if (node.Id == rootId.Id)
+            {
+                const auto dependencies = GetDependencies(manifest, DependencyType::Package);
+                if (!dependencies.size())
+                {
+                    return depList;
+                }
+
+                std::for_each(dependencies.begin(), dependencies.end(), [&](Dependency dep) { depList.Add(dep);  });
+                return depList;
+            }
+
+            auto packageLatest = GetPackageLatestVersion(index, node.Id);
+            if (!packageLatest.has_value())
+            {
+                std::string error = ManifestError::MissingManifestDependenciesNode;
+                error.append(" ").append(node.Id);
+                dependenciesError.emplace_back(ValidationError(error));
+                foundErrors = true;
+                return depList;
+            }
+
+            if (node.MinVersion > packageLatest.value().second)
+            {
+                std::string error = ManifestError::NoSuitableMinVersion;
+                error.append(" ").append(node.Id);
+                dependenciesError.emplace_back(ValidationError(error));
+                foundErrors = true;
+                return depList;
+            }
+
+            auto packageLatestDependencies = index->GetDependenciesByManifestRowId(packageLatest.value().first);
+            std::for_each(
+                packageLatestDependencies.begin(),
+                packageLatestDependencies.end(),
```

Comment:
range-for is much easier to read.

Reply:
I've had no contention so far with using collections related methods for any sort of data aggregation for elegance and to take advantage of whatever performance gain is associated with the underlying framework, I don't see why it should be any different in this case. Unless there is some sort of unwritten rule about avoiding them in this codebase, because I don't see any difference in readability between this and what is suggested, besides it's plastered almost everywhere in the dependencies implementation.

Reply:
There is nothing wrong with using `algorithm`, but it likely generates the exact same code.  It is just that the way you have it written:
```
std::for_each(
    container.begin(),
    container.end(),
    [&] (const auto& value)
    {
        <body>
    });
```
Is several more lines and one indentation deeper, containing a bunch of boilerplate that just adds noise.  The same thing with range-for is one line to indicate the loop:
```
for (const auto& value : container)
{
    <body>
}
```
I'm not going to reject your change solely on use of `for_each`, but I'm very comfortable with my stance on readability difference of those two. 😄